### PR TITLE
#867: add a view for comparing schema versions

### LINKED ISSUE #867: Ability to compare topic's schema across versions
### Is your proposal related to a problem?
As a user, I'd like to compare different versions of the specific schema (e.g. the latest with the previous ones)

(Write your answer here.)

### Describe the solution you'd like

(Describe your proposed solution here.)

### Describe alternatives you've considered

(Write your answer here.)

### Additional context

(Write your answer here.)

### PR DESCRIPTION
**Pull request by IndeedSi**
- [ ] **Breaking change?** (if so, please describe the impact and migration path for existing applications:)

**What changes did you make?** (Give an overview)
I added a view for comparing different schema versions. Fixes #867 

**Is there anything you'd like reviewers to focus on?**
Please check the UI

**How Has This Been Tested?** (put an "X" next to an item)

- [ ] No need to
- [ ] Manually(please, describe, when necessary)
- [ ] Unit checks
- [ ] Integration checks
- [ ] Covered by existing automation

**Checklist** (put an "X" next to an item, otherwise PR will fail)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (e.g. **ENVIRONMENT VARIABLES**)
- [x] My changes generate no new warnings(e.g. Sonar is happy)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged

Check out [Contributing](https://github.com/provectus/kafka-ui/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/provectus/kafka-ui/blob/master/CODE-OF-CONDUCT.md)

### CODE REVIEW

**Comment by IndeedSi** [comment]
Probably need to rebase when the redesign is merged :P

**Comment by Haarolean** [comment]
> Probably need to rebase when the redesign is merged :P

Yeah, that's why we were not planning any features for 0.3. 

@GneyHabub this can wait until the redesign is merged in. Up to you.

**Comment by Haarolean** [comment]
@IndeedSi could you rebase please?

**Review by workshur — CHANGES_REQUESTED** [review]
@NelyDavtyan pls rework it using our new styled-components approach

**Review by Motherships — COMMENTED** [review]
Request review when you will be done with changes requested by Oleg, please

File: kafka-ui-react-app/src/components/App.scss [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+          background-color: rgba(34,41,47,.5);
+          left: 0;
+          right: 0;
+          opacity: 1;
+          visibility: visible;
+        }
+      }
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (NelyDavtyan) [reply kafka-ui-react-app/src/components/App.scss:104]: @workshur should I delete this file?

File: kafka-ui-react-app/src/components/common/DiffViewer/__tests__/DiffViewer.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import { shallow } from 'enzyme';
```
Comment (Motherships) [inline kafka-ui-react-app/src/components/common/DiffViewer/__tests__/DiffViewer.spec.tsx:1]: Do we really need to have `enzyme` in new code when we have issue to get rid of it?

File: kafka-ui-react-app/src/components/Schemas/Diff/Diff.styled.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  flex-shrink: 1;
+  min-height: min-content;
+  padding-top: 1.5rem !important;
+  &
+    .ace_editor
+    > .ace_scroller
+    > .ace_content
+    > .ace_marker-layer
+    > .codeMarker {
+    background: ${Colors.yellow[20]};
```
Comment (Motherships) [inline kafka-ui-react-app/src/components/Schemas/Diff/Diff.styled.ts:18]: Direct use of `Colors`

File: kafka-ui-react-app/src/components/Schemas/Diff/Diff.tsx [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+  const { clusterName, subject } =
+    useParams<{ clusterName: string; subject: string }>();
+  const dispatch = useAppDispatch();
+
+  React.useEffect(() => {
+    dispatch(fetchSchemaVersions({ clusterName, subject }));
+    return () => {
+      dispatch(resetLoaderById(SCHEMAS_VERSIONS_FETCH_ACTION));
+    };
+  }, [clusterName, subject]);
```
Comment (Motherships) [inline kafka-ui-react-app/src/components/Schemas/Diff/Diff.tsx:47]: `react-hooks/exhaustive-deps` suggests to add `dispatch` to array of deps

File: kafka-ui-react-app/src/components/Schemas/Details/LatestVersionItem.tsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  schema: SchemaSubject;
+}
+
+const LatestVersionItem: React.FC<LatestVersionProps> = ({
+  schema: { id, subject, schema, compatibilityLevel },
+}) => (
+  <div className="tile is-ancestor mt-1">
+    <div className="tile is-4 is-parent">
+      <div className="tile is-child">
+        <table className="table is-fullwidth">
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Schemas/Details/LatestVersionItem.tsx:15]: we already have version without bulma classes.

pls use styled components from this view https://github.com/provectus/kafka-ui/blob/master/kafka-ui-react-app/src/components/Schemas/Details/LatestVersion/LatestVersionItem.tsx

File: kafka-ui-react-app/src/components/common/DiffViewer/__tests__/DiffViewer.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    renderComponent({ leftVersion: left, rightVersion: right });
+  });
+
+  it('renders with fixed height', () => {
+    renderComponent({
+      leftVersion: left,
+      rightVersion: right,
+      isFixedHeight: true,
+    });
+  });
```
Comment (workshur) [inline kafka-ui-react-app/src/components/common/DiffViewer/__tests__/DiffViewer.spec.tsx:32]: I believe we need to add at least one `expect` to each case

File: kafka-ui-react-app/src/components/Schemas/Details/Details.tsx [hunk 7]
```diff
@@ -83,6 +87,14 @@ const Details: React.FC = () => {
       <PageHeading text={schema.subject}>
         {!isReadOnly && (
           <>
+            <Button
+              isLink
+              buttonSize="M"
+              buttonType="primary"
+              to={clusterSchemaSchemaDiffPath(clusterName, subject)}
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Schemas/Details/Details.tsx:94]: where are the `leftVersion` and `rightVersion`?

File: kafka-ui-react-app/src/components/Schemas/Diff/Diff.tsx [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+
+  const methods = useForm({ mode: 'onChange' });
+  const {
+    formState: { isSubmitting },
+    control,
+  } = methods;
+
+  const searchParams = React.useMemo(
+    () => new URLSearchParams(location.search),
+    [location, history, leftVersion, rightVersion]
```
Comment (Motherships) [inline kafka-ui-react-app/src/components/Schemas/Diff/Diff.tsx:69]: I get why we have `location` in deps

Why we have `history, leftVersion, rightVersion` as deps?

**Review by Motherships — APPROVED** [review]

File: kafka-ui-react-app/src/components/common/DiffViewer/DiffViewer.tsx [hunk 9]
```diff
@@ -0,0 +1,52 @@
+/* eslint-disable react/jsx-props-no-spreading */
```
[the remark this answers is not in the document]
Comment (NelyDavtyan) [reply kafka-ui-react-app/src/components/common/DiffViewer/DiffViewer.tsx:1]: will remove it

File: kafka-ui-react-app/src/components/Schemas/Diff/Diff.tsx (hunk 8 above)
Comment (NelyDavtyan) [reply kafka-ui-react-app/src/components/Schemas/Diff/Diff.tsx:69]: you are right, we don't need them, will remove them

**Comment by Haarolean** [comment]
@IndeedSi thank you for your contribution!

**Pull request merged**